okonomiyaki/file_formats/egg.py

```diff
@@ -66,9 +67,21 @@ def add_file_as(self, path, archive_name):
         """ Add the given file to the egg, under the given archive name."""
         self._fp.write(path, archive_name)
 
-    def add_data(self, data, archive_name):
+    def add_data(self, data, archive_name, chmod=0o644):
         """ Write the given data as the given archive name."""
-        self._fp.writestr(archive_name, data)
+        if archive_name[-1] == "/":
+            raise ValueError(
+                "Invalid member name for data: '{}'".format(archive_name)
+            )
+
+        if not isinstance(archive_name, zipfile.ZipInfo):
+            zinfo = zipfile.ZipInfo(
+                filename=archive_name,
+                date_time=time.localtime(time.time())[:6]
+            )
+            zinfo.compress_type = self._fp.compression
+            zinfo.external_attr = chmod << 16
+        self._fp.writestr(zinfo, data)
```

Comment:
Will this fail if `isinstance(archive_name, zipfile.ZipInfo)` == `True` because `zinfo` will not be defined?